### DISCUSSION TITLE
How to add advice to achieve for org buffer like to collapse all other headings after executing 'consult-org-heading'?

### DISCUSSION THREAD

**Discussion by IceAsteroid** [minad/consult#943] [Q&A] [answered]
Hi, guys, thank you for the great effort!

I wish to have all other headings collapsed after selecting a heading via 'consult-org-heading'.

I wrote a snippet like follows, but it didn't seem to work

```
  (defun my/outline-collapse-all-other-headings
      (save-excursion
	(outline-hide-other)))
  (advice-add 'consult-org-heading :after 'my/outline-collapse-all-other-headings)
```

What's wrong?

**Comment by minad** [comment]
Hi! It could be that you have to use an Org-specific function, e.g., `org-fold-hide-sublevels` or some function from org-fold.el.

**Reply by IceAsteroid** [reply]
No, it doesn't work at all, it seems that such commands need to be executed while the org file buffer is focused. Otherwise, commands won't take effect.

I guess the mechanics of consult combined with vertico are different from other commands. I don't know what really has happened

**Reply by minad** [reply]
> No, it doesn't work at all, it seems that such commands need to be executed while the org file buffer is focused. Otherwise, commands won't take effect.

Right. But the after advice should be executed in the original window?

**Reply by IceAsteroid** [reply]
I'm still not good at elisp, the way I was doing might be the wrong way.

Could you help me on this, I'm sorry

The title was misleading, instead, I did mean to collapse all other headings, not all the headings, because there are cases when we select a sub-heading within its upper level headings.

**Reply by minad** [reply]
You could maybe collapse all headings in a :before advice, then after selection you will end up with only the selected heading open and the rest collapsed.

**Reply by IceAsteroid** [reply]
I've run the following code, but it still doesn't work

```
  (defun my/outline-collapse-all-headings ()
      (org-fold-hide-sublevels 1))
  (advice-add 'my/outline-collapse-all-headings :before 'consult-org-heading)
```

**Reply by minad** [reply]
Your advice is not correct. Please take a look at the documentation of `advice-add`. See https://github.com/minad/consult/discussions/943#discussioncomment-8596765 for the solution.

**Comment by IceAsteroid** [comment]
@minad, I'm still learning the basic elisp, could you suggest me a path to figure out what to do?

Maybe using advice is a wrong way, or I need to let it be combined with the use of some window programming like "select-window" to do the job? 

Actually, such a task does need some kind of a hook to easily do what it needs, but I don't see any of it.

**Reply by minad** [reply, marked as answer]
This works for me:

~~~ elisp
(advice-add #'consult-outline :before (lambda (&rest _) (org-fold-hide-sublevels 1)))
~~~

**Reply by IceAsteroid** [reply]
It works. Thank you @minad !!

By imitating your code, this one also works:
```
  (defun my/outline-collapse-all-headings ()
    ;;(org-fold-hide-sublevels 1)
    (outline-hide-sublevels 1)
    )
  (advice-add #'consult-outline :before (lambda (&rest _) (my/outline-collapse-all-headings)))
  (advice-add #'consult-outline :after (lambda (&rest _)
					 (outline-show-entry)
					 (outline-show-children)))
```

Format shown in GNU documents like this would not work, may be some arguments are passed to without knowing, so it needs to be reset for arguments
```
(advice-add #'consult-outline :before #'my/outline-collapse-all-headings)))
```
It shows ```Wrong number of arguments: (lambda nil (outline-hide-sublevels 1)), 1```

I've been figuring out recently, but it seems that's my problem to understand GNU's documents
